### ISSUE TITLE
wmvare-iso unable to start machine on windows - invalid path

### ISSUE THREAD

**Issue by fekir** [hashicorp/packer#5995]
I'm not able to test this on Linux, so I do not know if it is windows-specific.

Given following template
````
{
	"builders": [
		{
			"type": "vmware-iso",
			"vm_name": "{{ user `vm_name` }}",
			"guest_os_type": "{{user `guest_os_type`}}",
			"disk_size": "{{user `disk_size`}}",
			"output_directory": "out/vmware/{{user `vm_name`}}_base",
			"boot_wait": "2m",
			"headless": "{{user `headless`}}",
			"iso_checksum": "{{user `iso_checksum`}}",
			"iso_checksum_type": "{{user `iso_checksum_type`}}",
			"iso_urls": [
				"{{user `iso_url`}}"
			]
			"communicator": "winrm",
			"winrm_password": "{{user `password`}}",
			"winrm_username": "{{user `username`}}",
			"winrm_timeout": "12h"
		}
	],
	"provisioners": [
	],
	"variables": {
		"cpus": "2",
		"disk_size": "65536",
		"memory": "2048",
		"vram": "32",
		"headless": "false",
		"iso_checksum": "ea214ee684a5bb8230707104c54a3b74d92f1d69",
		"iso_checksum_type": "sha1",
		"iso_url": "./iso/en_windows_10_multi-edition_version_1709_updated_dec_2017_x64_dvd_100406711.iso",
		"password": "vagrant",
		"username": "vagrant",
		"version": "0.1.0",
		"vm_name": "windows",
		"guest_os_type": "windows9-64"
	}
}
````

vmware hangs during the boot process, ie I can just see the blue windows logo on a black background. 
vmware does not freeze, but the machine does not boot up/the installation process of windows does not start.

Further informations:

vmware `11.1.0 build-2496824`
Host: `Windows 10, 64-bit  (Build 16299) 10.0.16299`
packer: `1.2.1`

output of `ll`:

```
total 8.4M
-rwxrwx---+ 1 XXX 1.1K Mar  8 08:33 disk.vmdk
-rwxrwx---+ 1 XXX 576K Mar  8 08:44 disk-s001.vmdk
-rwxrwx---+ 1 XXX 512K Mar  8 08:33 disk-s002.vmdk
-rwxrwx---+ 1 XXX 512K Mar  8 08:33 disk-s003.vmdk
-rwxrwx---+ 1 XXX 512K Mar  8 08:33 disk-s004.vmdk
-rwxrwx---+ 1 XXX 512K Mar  8 08:33 disk-s005.vmdk
-rwxrwx---+ 1 XXX 512K Mar  8 08:33 disk-s006.vmdk
-rwxrwx---+ 1 XXX 512K Mar  8 08:33 disk-s007.vmdk
-rwxrwx---+ 1 XXX 512K Mar  8 08:33 disk-s008.vmdk
-rwxrwx---+ 1 XXX 512K Mar  8 08:33 disk-s009.vmdk
-rwxrwx---+ 1 XXX 512K Mar  8 08:33 disk-s010.vmdk
-rwxrwx---+ 1 XXX 512K Mar  8 08:33 disk-s011.vmdk
-rwxrwx---+ 1 XXX 512K Mar  8 08:33 disk-s012.vmdk
-rwxrwx---+ 1 XXX 512K Mar  8 08:33 disk-s013.vmdk
-rwxrwx---+ 1 XXX 512K Mar  8 08:33 disk-s014.vmdk
-rwxrwx---+ 1 XXX 512K Mar  8 08:33 disk-s015.vmdk
-rwxrwx---+ 1 XXX 512K Mar  8 08:33 disk-s016.vmdk
-rwxrwx---+ 1 XXX 128K Mar  8 08:33 disk-s017.vmdk
-rwxrwx---+ 1 XXX 190K Mar  8 08:44 vmware.log
-rwxrwx---+ 1 XXX 8.5K Mar  8 08:44 windows.nvram
-rwxrwx---+ 1 XXX    0 Mar  8 08:33 windows.vmsd
-rwxrwx---+ 1 XXX 3.9K Mar  8 08:44 windows.vmx
-rwxrwx---+ 1 XXX  262 Mar  8 08:33 windows.vmxf
```

log files and generated vmx file:
[packer.log](https://github.com/hashicorp/packer/files/1794323/packer.log)
[vmware.log](https://github.com/hashicorp/packer/files/1794324/vmware.log)
[windows.vmx.txt](https://github.com/hashicorp/packer/files/1794326/windows.vmx.txt)

I was not able to see anything suspicious. 

When clicking on the properties of the VM, and then on the virtual disk (_NOT the ISO file_), I get an error (see image).
![error_when_clicking_on_virtualdisk](https://user-images.githubusercontent.com/11912465/37169051-19c65de0-2307-11e8-8bc8-a901b156bfc4.png)

Unfortunately I cannot see any of the mentioned characters in any configuration file, or in vmware working directory (there are also no spaces at the begin or end of any filename).

Manually creating and starting a virtual machine with the same iso works without issues.

I waited until the line "--- at this point i triggered the shutdown" (added manually in the `vmware.log` file).

Trying to start the machine manually (after copying the directory) does not make any difference.

**Comment by SwampDragons** [comment]
Forgive me; my windows-fu is weak. It seems to me that you're running vmware workstation from the windows linux subsystem (going off your ll output, which looks like bash rather than like cmd or powershell).  Perhaps it's the drive name `C:` messing things up? What if you use the linux pathing instead (`/mnt/c/INIMA/iso/...`)

**Comment by fekir** [comment]
Hi, the `ll` output is from cygwin, just to show that the files where created, but I did run packer from a normal powershell console. I did also test it from cmd with the same result, sorry forgot to mention that.
I used cygwin to ensure that there where, for example, no whitespace characters at the begin or end of the name, since in those cases the names would have been quoted.

The gui from vmware (where I took the screenshot) was started from the shortcut.

I do not have WSL on my machine.

**Comment by SwampDragons** [comment]
Ah, @fekir I didn't look at your config closely enough, and I think this iso-filename error is not the real problem. You don't seem to be passing in any [floppy_files](https://www.packer.io/docs/builders/vmware-iso.html#floppy_files). In order to allow Packer to connect via WinRM, you need to use floppy_files to to open the vm's firewall and launch WinRM on the guest VM. 

We have an example of a very minimal file [here](https://www.packer.io/intro/getting-started/build-image.html#a-windows-example); I know that example is for AWS, but you should be able to save the file and pass it in as a floppy file.

Or you can look at some community examples: https://github.com/joefitzgerald/packer-windows/blob/master/windows_10.json

I'm going to close this issue since it seems to be more of an initial setup question than a problem with Packer; if you still have trouble launching your first instance, reach out to the mailing list [here](https://groups.google.com/forum/#!forum/packer-tool) which will get you more help from people with more experience with your exact setup.

**Comment by fekir** [comment]
Hi, would it be possible to reopen it?

I've removed the floppy file in order to provide a minimal example.
My first attempt was with a floppy file for Autounattanded.xml, and it did not work with the same error message.
And if not floppy file is present, the expected behavior would be that the virtual machine starts the iso, like when creating the vm manually.

I already looked at the samples, if I use virtualbox it works without problems.

**Comment by SwampDragons** [comment]
I see. But you do get a boot screen. So it sounds like the iso is being started.

**Comment by fekir** [comment]
![boot](https://user-images.githubusercontent.com/11912465/37786359-ce5dc87e-2dfc-11e8-98c0-d9963bfd41b6.png)
This is all I get: the blue logo on a completely black screen.
It seems to get started, but it does not do anything, there is not even the white circle progress bar.
There is no wizard for starting the installation or anything.

And there is also that cryptic error message about the filename when I'm inspecting the vm properties.

I did not mention it, but just to be super-sure that there were not strange characters in my path either, I tested every in a directory directly under `C:`.

**Comment by SwampDragons** [comment]
Thanks. [packer.zip](https://github.com/hashicorp/packer/files/1838594/packer.zip)

**Comment by SwampDragons** [comment]
Has this ever worked for you with any version of Packer?

**Comment by fekir** [comment]
> Thanks. packer.zip
Same behavior as before

> Has this ever worked for you with any version of Packer?
I did not test with previous versions of packer, so I do not know.
I thought that vmware was a supported platform, so I expected it to work.

> are the vmdk's in that output directory old?
The directories have been fully created by packer.

I've also tested with a debian iso (and template), and got a similar result (blue screen, but not at the beginning, only after the ip settings where configured).

I tried to remove and reattach the virtual disk from vmware gui, and this is what I got:
![bug_add_disk_again](https://user-images.githubusercontent.com/11912465/37840002-414c2104-2ebc-11e8-98a5-85a7173f9ce3.png)

It seems that the virtual disk created by packer are not compatible with this version of vmware.
I have no idea why there is no warning in vmware logs, but I would therefore expect that this issue is not windows specific, as I thought at the beginning.

It also means that in both cases (windows and debian) the setup starts correctly, but since there is no disk attached it simply does nothing (seems pretty dumb, an error message would probably ahev helped).
Also vmware interface was not very consistent, since the virtual disk is showed as attached.

**Comment by arizvisa** [comment]
@fekir, actually packer doesn't create any virtual disks. It lets VMware handle all of that by using vmrun.exe which comes with VMware Workstation. I tested opening up the .vmx you provided and I do not seem to have this issue.

But also, I'm using 14.0.0 build-6661328.

If you open up the .vmx in an editor and change the path to your iso so that all forward-slashes are back-slashes, does that fix the issue?

**Comment by arizvisa** [comment]
@fekir, sorry forgot to include the line that includes the path to the .iso.
In that .vmx you included, it's at `ide0:0.filename = "C:\INIMA\iso..."`
Try swapping the backslashes for forward-slashes or seeing if that makes a difference.

Your screenshot shows forward-slashes, but your .vmx shows back-slashes which is where my confusion was at.

**Comment by fekir** [comment]
Unfortunately the error message of vmware is completely misleading, it seems it has nothing to do with the iso file.
Otherwise the windows logo would not appear and the Debian installer would not start.

In any case I changed the path in the `.vmx` file as you proposed, but nothing changed.

I also made another experiment:
When creating manually (a VM), I can choose the compatibility level.
If I set it to version 9 (the same version used by packer), and create a new machine, I'm unable to add the virtual disk created with packer, and get exactly the same message described in https://user-images.githubusercontent.com/11912465/37840002-414c2104-2ebc-11e8-98a5-85a7173f9ce3.png

On the other hand if I create a new virtual machine with version 11, then i'm able to add without issues the `.vmdk` created by packer (yes, not packer but hold with me: the virtual disk created by vmware wen running it with packer).

@arizvisa: I meant the iso created by packer since packer does not tell vmware to create a virtual machine, but creates a virtual disk separately.
`Executing: C:/Program Files (x86)/VMware/VMware Workstation/vmware-vdiskmanager.exe -c -s 65536M -a lsilogic -t 1 out\vmware\windows_base\disk.vmdk`

It looks like the `vmdk` generated by this command is not compatible with the generated virtual machine, probably somewhere it is necessary to tell vmware that the virtual machine should be a Wokstation 11.0 or something like that.

**Comment by arizvisa** [comment]
So, changing the path in .vmx still displays the error message, "`The characters > < * : / * ? and | are not allowed in the ISO image file.`" like in your screenshot?

**Comment by arizvisa** [comment]
Is your issue in the iso, or is it in the .vmdk?? The screenshots also show that there's a missing floppy disk.

But, if it's in the .vmdk compatibility, then @SwampDragons was right to close the issue as you created it while discussing an issue about not being able to boot up off the iso. If it's in the operating system not being able to detect any hard disks, then this is a completely different issue.

**Comment by fekir** [comment]
@arizvisa: Let me sum up what the problem is:

The virtual machine created by packer is not usable for installing an operating system.
I'm not 100% sure what the cause is, I'm just reporting my find-outs.

> So, changing the path in .vmx still displays the error message, "The characters > < * : / * ? and | are not allowed in the ISO image file." like in your screenshot?

Yes, and the vm starts but the installation does not begin.

> Is your issue in the iso, or is it in the .vmdk?? The screenshots also show that there's a missing floppy disk.

From the error message it seems to be the path of the iso, but since I can remove and re-add the iso file without problems, but can't remove and re-add the vmdk file, I guess it's the vmdk.
The error message also appears when _clicking on the vmdk_ and not the iso (I did report this immediately).
Some things I did not realize immediately, and reported them afterwards.
Also notice that I made afterwards the same test with Debian (I mean I tried to install debian).
I have the exact same behavior, expect for the fact that the Debian installations begin, but when trying to identify the disk where to install it simply stops with a blue screen. (Only after this finding I tried to remove and re-attach the vmdk file).
This is also a hint that the iso file is completely fine.

There is no floppy disk because I wanted to provide a minimal sample, the windows iso does not need a floppy file for starting correctly.

Now, since you said you where able to use my vmx file (how did you without the vmdk files?) on VMWare 14, I just downloaded on a Ubuntu machine the trial version of VMWare (version 14).
I get the exact same behavior with all issues described here, the only difference is that in the options dialog, when clicking on the vmdk settings, I do not get this error: https://user-images.githubusercontent.com/11912465/37169051-19c65de0-2307-11e8-8bc8-a901b156bfc4.png

But everything else is the same:
 - The installation process of windows and Debian hangs
 - I'm unable to remove and re-attach the vmdk file with the error message that the disk is incompatible with Workstation 9.
 - When creating a virtual machine by hand, it creates a Workstation 14 compatible machine, and in that machine I'm able to attach the vmdk files created by packer. When creating manually a Workstation 9 machine, I get the same compatibility error when trying to attach the vmdk files created by vmware.

So it is not only a problem on Windows with wmvare 11, but also on Ubuntu 16 (LTS) with vmware 14.

And just to make clear: It is not an issue with my template, I also tried fitzgeralds and others as suggested by  SwampDragons, and I get exactly the same behavior (I'm unable to test parallels since I do not have a Mac where I can play).


The issue is not about being unable to boot up the iso, the issue is that I'm not able to use packer with vmware (see the title: wmvare-iso unable to start machine on windows. OK, in the meantime we know that the machine starts, but it does not go any further), similar to https://github.com/hashicorp/packer/issues/5969 with QEMU and at the beginning, from wmvare error message, it seemed that the issue was the iso *path*, even if I was 100% sure that the paths where fine.

**Comment by arizvisa** [comment]
You can create a .vmdk with vmware-vdiskmanager, but you knew that already.
Because it's listed in your packer log: `C:/Program Files (x86)/VMware/VMware Workstation/vmware-vdiskmanager.exe -c -s 65536M -a lsilogic -t 1 out\vmware\windows_base\disk.vmdk`

Really, it'll be best to create a new issue as the this specific issue first mentions a problem with the machine starting, then the .iso path having invalid characters, and then the .vmdk disk not being compatible.

Both of these insights actually have nothing to do with the OS actually booting up and stalling as the .iso path still results in your instance of Windows booting up, and the .vmdk disk isn't even touched because the OS installer hasn't even kicked in yet. Thus troubleshooting both of these symptoms (although they are symptoms), will lead down the wrong path which is why it's been 21 days, and you're still in the same place.

Really the question/issue that you're asking is why is the operating system stalling whenever you use packer. This problem is completely contrary to what you've described in this issue because the machine does actually boot up, the operating system just hangs.

The way of troubleshooting this would likely be to test out a different operating system where you can watch the boot sequence progress. I think that even packer's acceptance tests might do this.

**Comment by arizvisa** [comment]
It's understandable that you might not feel like creating an issue because you created a commit that points at this id#, but you need to consider other users whom might be searching tickets for a similar problem. Since you described this as a machine powering-on problem related to an invalid path, a user will only be able to locate this issue when they have a problem with those specific characteristics.

Since the problem has been troubleshot past the point of the machine not booting up and is now at a point where it's the operating system that is stalling, it's now a completely different problem which is why the issue was originally closed.

So, the next way to proceed is to troubleshoot why the operating system is stalling since the invalid paths seem to have nothing to do with the machine starting or not

**Comment by fekir** [comment]
> Really, it'll be best to create a new issue as the this specific issue first mentions a problem with the machine starting, then the .iso path having invalid characters, and then the .vmdk disk not being compatible.
Those where the findouts when troubleshooting the problem. But I have nothing against creating a new issue. Ignore the commit, its just a placeholder for me, and will get deleted when I'll be able to test my configurations.

> this specific issue first mentions a problem with the machine starting, then the .iso path having invalid characters, and then the .vmdk disk not being compatible.

I think there might be some misunderstanding, or my description was not very clear.
I reported since the beginning that I was getting the error message when clicking on the **virtual disk** and **not the iso** file.
I also added that the iso file is fine since I can use it without issues on other machines.
I never wrote that the issue was in the iso, just reported the error message.

> Really the question/issue that you're asking is why is the operating system stalling whenever you use packer. 
And that's the question I'm asking myself from the beginning.
Yes, I think the issue is the vmdk file, but it could also something else.
But since I can reproduce it on two completely unrelated systems, I cannot imagine why you cannot reproduce it.

> The way of troubleshooting this would likely be to test out a different operating system where you can watch the boot sequence progress. I think that even packer's acceptance tests might do this.

Well, as reported I already tested Windows and Debian guest on Windows and Ubuntu host. Same behavior.


As far as I can see the only difference with the initial report it's that it is not vmware that hangs, but the guest OS installer. Everything that comes later are additional information.
If you still think that it might be better to create a new issue I have nothing against it.

Should I create it?
If yes, which information should I copy from here?
I would copy everything, but then it would make little sense.

**Comment by fekir** [comment]
I fail to see why it would be a networking issue, since 
- starting the created machine manually does not work either
- removing and attaching the vmdk produces an error

Could you give me some insight? Or resource where I could look at?
Maybe I can gather some information more.

**Comment by fekir** [comment]
I swear I was going to open a separate ticket.

I spent half a day testing different configurations, (couple of windows iso, network enabled/disabled and so on, in order to provide as much informations as possible and to ensure that I did not make any dumb mistake), and at last I made one final change.

In VMware 14 there is a nice button to upgrade the virtual machine to it's last format.
I upgraded it the machine from the old version and the setup started!
So i changed the virtualhw.version to version 11 or never: and the packer template worked.

The most strange thing is that I did not found any project thats sets this variable.
For example, https://github.com/joefitzgerald/packer-windows/blob/master/windows_10.json (since it was mentioned by SwampDragons) sets some vmx_data, but not this one, not even in the vagrant configuration files.
And since the virtualhw.version value is not equal to the installed version, I have no idea how such templates can work on other machines and why you were not able to reproduce the problem.

Unfortunately I do not have access anymore the the machine with VMware 11, so I could only test it with VMware 14, but I think that this would solve the problem even on VMware 11.

Since now I can create vmware machines, I see little reason opening a new ticket, and I think this can also get closed too.

I just have one last question: is it possible to set the version to the latest automatically?
From https://github.com/hashicorp/packer/issues/1191 it seems that it cannot be done.

Side note: if I set virtualhw.version to the highest possible version (14 in this case), I do not get any errors anymore when removing and the adding the vmdk created by packer.
